nucliadb/src/nucliadb/search/search/rerankers.py

```diff
@@ -187,3 +145,58 @@ def get_reranker(kind: search_models.Reranker) -> Reranker:
         logger.warning(f"Unknown reranker requested: {kind}. Using multi-match booster instead")
         reranker = MultiMatchBoosterReranker()
     return reranker
+
+
+def apply_reranking(results: KnowledgeboxFindResults, reranked: list[RankedItem]):
+    inverted_results = {}
+    for rid, resource in results.resources.items():
+        for field_id, field in resource.fields.items():
+            for paragraph_id, paragraph in field.paragraphs.items():
+                inverted_results[paragraph_id] = (
+                    paragraph,
+                    (field_id, field),
+                    (rid, resource),
+                )
+
+    # order results by new score
+    best_matches = []
+    for item in reranked:
+        paragraph = inverted_results[item.id][0]
+        paragraph.score = item.score
+        paragraph.score_type = item.score_type
+        best_matches.append((item.id, item.score))
+
+    best_matches.sort(key=lambda x: x[1], reverse=True)
+
+    # update best matches according to new scores
+    cut = results.page_size
+    results.best_matches.clear()
+    for order, (paragraph_id, _) in enumerate(best_matches[:cut]):
+        paragraph = inverted_results[paragraph_id][0]
+        paragraph.order = order
+        results.best_matches.append(paragraph_id)
+
+    # cut uneeded results
+    extra = best_matches[cut:]
+    for paragraph_id, _ in extra:
+        _, (field_id, field), (rid, resource) = inverted_results[paragraph_id]
+        field.paragraphs.pop(paragraph_id)
+        if len(field.paragraphs) == 0:
+            resource.fields.pop(field_id)
+
+        if len(resource.fields) == 0:
+            results.resources.pop(rid)
+
+
+def _get_items_to_rerank(results: KnowledgeboxFindResults) -> list[RerankableItem]:
+    return [
+        RerankableItem(
+            id=paragraph_id,
+            score=paragraph.score,
+            score_type=paragraph.score_type,
+            content=paragraph.text,
+        )
+        for rid, resource in results.resources.items()
+        for field_id, field in resource.fields.items()
```

Comment:
nit: these two could be values

```suggestion
        for resource in results.resources.values()
        for field in resource.fields.values()
```